### PR TITLE
Migrate build logic to conventions plugins

### PR DESCRIPTION
**Pull request by haydenbaker**
#### Background
* Overhauls our builds to use convention plugins, which is a far cleaner approach
* Spotless and formatting will be added/applied in a follow-up PR (separate to keep this PR from being several pages)

#### Testing
* `./gradlew clean build publish -c -Plog-tests`
* tested jreleaser configs (for main and cli release)
* checked shadowed publications, verified shadowed deps are not in the pom

---
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: buildSrc/src/main/kotlin/Extensions.kt [hunk 1]
```diff
@@ -0,0 +1,11 @@
+import org.gradle.api.Project
+import org.gradle.api.file.Directory
+import org.gradle.api.provider.Provider
+import org.gradle.kotlin.dsl.*
```
Comment (yasmewad) [inline buildSrc/src/main/kotlin/Extensions.kt:4]: nit: add single line imports?

File: build.gradle.kts [hunk 2]
```diff
@@ -13,296 +16,50 @@
...
-    `maven-publish`
-    signing
-    checkstyle
-    jacoco
-    id("com.github.spotbugs") version "6.0.8"
-    id("io.codearte.nexus-staging") version "0.30.0"
-    id("me.champeau.jmh") version "0.7.2"
-    id("com.github.johnrengelman.shadow") version "7.1.2"
-    id("org.jreleaser") version "1.12.0" apply false
+    id("org.jreleaser") version "1.15.0" apply false
```
Comment (hpmellema) [inline build.gradle.kts:20]: could we add an alias here and move this into version library? Having everything in one place makes the dependabot changes tiny.

File: build.gradle.kts [hunk 3]
```diff
@@ -13,296 +16,50 @@
...
-    }
-}
-
-// The root project doesn't produce a JAR.
-tasks.named("jar") {
-    enabled = false
-}
-
-// ==== Javadoc ====
-afterEvaluate {
```
Comment (hpmellema) [inline build.gradle.kts:253]: I dont see any aggregate javadoc task anymore? Maybe I missed it, but we still need an aggregate javadoc task.

File: buildSrc/src/main/kotlin/smithy.module-conventions.gradle.kts [hunk 4]
```diff
@@ -0,0 +1,5 @@
+plugins {
+    id("smithy.java-conventions")
+    id("smithy.integ-test-conventions")
```
Comment (hpmellema) [inline buildSrc/src/main/kotlin/smithy.module-conventions.gradle.kts:3]: I dont think all modules need integ test conventions. I would prefer to isolate this to the 2 instances that need this.

File: smithy-cli/build.gradle.kts [hunk 5]
```diff
@@ -0,0 +1,367 @@
...
+import org.gradle.api.tasks.testing.logging.TestExceptionFormat
+import org.gradle.api.tasks.testing.logging.TestLogEvent
+import org.jreleaser.gradle.plugin.JReleaserExtension
+import org.jreleaser.model.Active
+import org.jreleaser.model.Distribution
+import org.jreleaser.model.Stereotype
+
+plugins {
+    application
+    id("org.beryx.runtime") version "1.13.1"
```
Comment (hpmellema) [inline smithy-cli/build.gradle.kts:16]: as with jreleaser, can we move these into version lib?

**Review by hpmellema — CHANGES_REQUESTED** [review]

**Review by hpmellema — APPROVED** [review]

**Pull request merged**